tasks/migration/migrate_system_package_data.go

```diff
@@ -0,0 +1,204 @@
+								Where("installable_id IS NULL AND applicable_id IS NULL").
+								Save(models.SystemPackage{
+									RhAccountID:   as.RhAccountID,
+									SystemID:      as.SystemID,
+									PackageID:     u.PackageID,
+									NameID:        u.NameID,
+									InstallableID: &installableID,
+									ApplicableID:  &applicableID,
+								}).Error
+						})
+						if err != nil {
+							utils.LogWarn("#", i, "Failed to update system_package2")
+						}
+					}
+				}
+			}(as, i, part)
+		}
+		wg.Wait()
+		utils.LogInfo("#", i, "partition", part, "Partition migrated")
+	}
+}
+
+func getAccSys(part string, i int) []AccSys {
+	// get systems from system_package partition
+	accSys := make([]AccSys, 0)
+	err := tasks.WithReadReplicaTx(func(db *gorm.DB) error {
+		return db.Table(part).
+			Distinct("rh_account_id, system_id").
+			Order("rh_account_id").
+			Order("system_id").
+			Find(&accSys).Error
+	})
+	if err != nil {
+		utils.LogWarn("#", i, "partition", part, "Failed to load data from partition")
+		return accSys
+	}
+
+	utils.LogInfo("#", i, "partition", part, "count", len(accSys), "Migrating systems")
+	return accSys
+}
+
+func getUpdates(as AccSys, part string, i int) []SystemPackageRecord {
+	var updates []SystemPackageRecord
+
+	// get update_data from system_package for given system
+	err := tasks.WithReadReplicaTx(func(db *gorm.DB) error {
+		return db.Table(part).
+			Select("name_id, package_id, update_data").
+			Where("rh_account_id = ?", as.RhAccountID).
+			Where("system_id = ?", as.SystemID).
+			Find(&updates).Error
+	})
+	if err != nil {
+		utils.LogWarn("#", i, "partition", part, "rh_account_id", as.RhAccountID, "system_id", as.SystemID,
+			"err", err.Error(), "Couldn't get update_data")
+	}
+	return updates
+}
+
+func getUpdateData(u SystemPackageRecord, as AccSys, part string, i int) []UpdateData {
+	var updateData []UpdateData
+	if err := json.Unmarshal(u.UpdateDataJSON, &updateData); err != nil {
+		utils.LogWarn("#", i, "partition", part, "rh_account_id", as.RhAccountID, "system_id", as.SystemID,
+			"update_data", string(u.UpdateDataJSON),
+			"err", err.Error(), "Couldn't unmarshal update_data")
+	}
+	return updateData
+}
+
+func getEvraApplicability(udpateData []UpdateData) (string, string) {
+	// get latest applicable and installable evra
+	var latestInstallable, latestApplicable string
+	for i := len(udpateData) - 1; i >= 0; i-- {
+		if len(latestInstallable) > 0 && len(latestApplicable) > 0 {
+			break
+		}
+		evra := udpateData[i].Evra
+		switch udpateData[i].Status {
+		case "Installable":
+			if len(latestInstallable) == 0 {
+				latestInstallable = evra
+			}
+			if len(latestApplicable) == 0 {
+				latestApplicable = evra
+			}
+		case "Applicable":
+			if len(latestApplicable) == 0 {
+				latestApplicable = evra
+			}
+		}
+	}
+
+	return latestApplicable, latestInstallable
+}
+
+func getPackageIDs(u SystemPackageRecord, i int, latestApplicable, latestInstallable string) (int64, int64) {
```

Comment:
Given the fact that it will query same EVRAs over and over - can't we read them all into map (memory) at the start of the job?

Reply:
We can try to re-use the same cache as we are using in evaluator, sure

Reply:
I;m not sure how much memory we will need so I set the limit to 1024Mi